Provide TypeScript code for every JavaScript code in the documentation

Muhammad-Altabba opened the pull request:
## Description

Fixes #6548

Additionally:
- Updated `docusaurus` and `react` dependencies in `docs`.
- Added a section to docs/READND.md file for how to add tabs for JavaScript/TypeScript
- Styled the tabs in harmony with the official logo colors of JavaScript/TypeScript
- Fixed and updated few code samples.
- However, the codes at Plugin Developers and migration guides kept in Typescript only.

## Type of change

<!-- Please delete options that are not relevant. -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] I have selected the correct base branch.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [ ] I ran `npm run lint` with success and extended the tests and types if necessary.
- [ ] I ran `npm run test:unit` with success.
- [ ] I ran `npm run test:coverage` and my test cases cover all the lines and branches of the added code.
- [ ] I ran `npm run build` and tested `dist/web3.min.js` in a browser.
- [ ] I have tested my code on the live network.
- [ ] I have checked the Deploy Preview and it looks correct.
- [ ] I have updated the `CHANGELOG.md` file in the root folder.
- [ ] I have linked Issue(s) with this PR in "Linked Issues" menu.

Muhammad-Altabba committed:
try commenting cypress special config for firefox

Muhammad-Altabba committed:
delete cypress firefox special configurations

Muhammad-Altabba committed:
update dependencies

Muhammad-Altabba committed:
add styles for javascript & typescript tabs

Muhammad-Altabba committed:
add taps snippet to README.md at docs

Muhammad-Altabba committed:
update smart_contracts docs

Muhammad-Altabba committed:
update and fix some codes at docs

Muhammad-Altabba committed:
add js/ts tabs at providers guide

Muhammad-Altabba committed:
add js/ts tabs at events listening and providers examples guides

Muhammad-Altabba committed:
add js/ts tabs at basic interaction docs

Muhammad-Altabba committed:
add js/ts tabs at  docs for advanced features

Muhammad-Altabba committed:
add js/ts tabs at docs for plugin users and migration guide

Muhammad-Altabba committed:
Merge into 6548-provide-typescript-code-for-every-javascript-code-in-the-documentation

Muhammad-Altabba committed:
Merge branch '4.x' into 6548-provide-typescript-code-for-every-javascript-code-in-the-documentation

avkos reviewed (changes requested):
That's amazing! I suggest some improvements:
1. Could we make a TypeScript tab a default one?
2. Could we change this border around the code? it does not very good

<img width="997" alt="image" src="https://github.com/web3/web3.js/assets/9366141/2a3c08cb-a18b-4e54-a5e2-a485b47c1b81">

For example like this:

<img width="917" alt="image" src="https://github.com/web3/web3.js/assets/9366141/124b9deb-5a2e-47d8-af72-fafd0a128e50">

Muhammad-Altabba committed:
Merge branch '4.x' into 6548-provide-typescript-code-for-every-javascript-code-in-the-documentation

jdevcs commented on docs/README.md:
```diff
@@ -39,3 +39,49 @@ $ GIT_USER=<Your GitHub username> yarn deploy
+
+### How to add tabs for Programming Languages
+
+Att the top of the `.md` file, ensure having the following: 
```
```suggestion
At the top of the `.md` file, ensure having the following: 
```
typo

jdevcs commented on docs/README.md:
```diff
@@ -39,3 +39,49 @@ $ GIT_USER=<Your GitHub username> yarn deploy
+</pre>
+
+Notes: 
+- This will put the JavaScript tab as the default. This is chosen because a JavaScript project needs less configurations. However, if the tutorial steps involves installing and configuring TypeScript, put the `default` attribute on the TypeScript tab instead.
```
```suggestion
- This will put the JavaScript tab as the default. This is chosen because a JavaScript project needs less configurations. However, if the tutorial steps involve installing and configuring TypeScript, put the `default` attribute on the TypeScript tab instead.
```

jdevcs commented on docs/README.md:
```diff
@@ -39,3 +39,49 @@ $ GIT_USER=<Your GitHub username> yarn deploy
+- This will put the JavaScript tab as the default. This is chosen because a JavaScript project needs less configurations. However, if the tutorial steps involves installing and configuring TypeScript, put the `default` attribute on the TypeScript tab instead.
+- The attribute `groupId="prog-lang"` ensure that whenever the user press on a tab, all other tabs in the page will switch to the same selected tab.
+- The attribute `queryString` is used to let the user easily share with the selected tab preserved.
+- The attributes `attributes={{className: "javascript-tab"}}` and `attributes={{className: "typescript-tab"}}` will give the tabes a unified design that uses the official colors of JavaScript and TypeScript.
```
```suggestion
- The attributes `attributes={{className: "javascript-tab"}}` and `attributes={{className: "typescript-tab"}}` will give the tabs a unified design that uses the official colors of JavaScript and TypeScript.
```

jdevcs commented on docs/docs/guides/advanced/support_additional_rpc_methods/index.md:
````diff
@@ -23,7 +26,27 @@ Following tutorial will guide you through the process of creating a custom plugi
+  <TabItem value="javascript" label="JavaScript"
+  	attributes={{className: "javascript-tab"}}>
+
+```javascript
````
pls set ts code as first in all examples

jdevcs commented on docs/docs/guides/advanced/support_additional_rpc_methods/index.md:
````diff
@@ -23,7 +26,27 @@ Following tutorial will guide you through the process of creating a custom plugi
+```javascript
+const { Web3PluginBase } = require('web3');
+
+export default class CustomRpcMethodsPlugin extends Web3PluginBase {
````
This js code is mixture of es and cjs and it will result in error!

using it CJS , there is :
```
export class CustomRpcMethodsPlugin extends Web3PluginBase {
^^^^^^

SyntaxError: Unexpected token 'export'
```
using esm:
```
const { Web3PluginBase } = require('web3');
                           ^
ReferenceError: require is not defined in ES module scope, you can use import instead
```

Muhammad-Altabba replied:
Sorry for the mess. It is corrected now.

jdevcs commented on docs/docs/guides/advanced/support_additional_rpc_methods/index.md:
````diff
@@ -32,19 +55,71 @@ export default class CustomRpcMethodsPlugin extends Web3PluginBase {
+```javascript
+const { Web3PluginBase } = require('web3');
+
+export default class CustomRpcMethodsPlugin extends Web3PluginBase {
````
this will also have error

jdevcs commented on docs/docs/guides/advanced/support_additional_rpc_methods/index.md:
````diff
@@ -32,19 +55,71 @@ export default class CustomRpcMethodsPlugin extends Web3PluginBase {
+```javascript
+const { Web3PluginBase } = require('web3');
+
+export default class CustomRpcMethodsPlugin extends Web3PluginBase {
````
same here

jdevcs commented on docs/docs/guides/advanced/support_additional_rpc_methods/index.md:
````diff
@@ -61,9 +136,38 @@ export default class CustomRpcMethodsPlugin extends Web3PluginBase {
+```javascript
+const { Web3PluginBase } = require('web3');
+
+export default class CustomRpcMethodsPlugin extends Web3PluginBase {
````
pls compile all run all code examples on your local system for this PR.

jdevcs replied:
pls compile & run all code examples on your local system for this PR.

jdevcs commented on docs/docs/guides/advanced/support_additional_rpc_methods/index.md:
```diff
@@ -61,9 +136,38 @@ export default class CustomRpcMethodsPlugin extends Web3PluginBase {
+const { Web3PluginBase } = require('web3');
+
+export default class CustomRpcMethodsPlugin extends Web3PluginBase {
+	public pluginNamespace = 'customRpcMethods';
```
The 'public' modifier can only be used in TypeScript files.

Muhammad-Altabba committed:
Merge branch '4.x' into 6548-provide-typescript-code-for-every-javascript-code-in-the-documentation

Muhammad-Altabba committed:
Apply suggestions from code review

Co-authored-by: Junaid <86780488+jdevcs@users.noreply.github.com>

Muhammad-Altabba committed:
remove border around code blocks in tabs

Muhammad-Altabba committed:
default code tab to typescript
except when there is a step by step guide with no steps to configure for typescript

Muhammad-Altabba committed:
fix javascript code samples at some docs

Muhammad-Altabba committed:
tiny enhancements at some docs

Muhammad-Altabba committed:
add a line to a doc

Muhammad-Altabba committed:
Merge branch '4.x' into 6548-provide-typescript-code-for-every-javascript-code-in-the-documentation

jdevcs reviewed (approved):
I compiled some of examples but not all in first review,
 @Muhammad-Altabba as discussed, before merging  check that all code examples should be compiling / running error free. thanks

avkos reviewed (approved):
LGTM. It will be great to have the Typescript tab first and selected in all examples.
